### PR TITLE
feat(plugin-lighthouse): add lighthouse plugin

### LINKED ISSUE #90: Implement Lighthouse plugin
> As a user,
> I want a plugin that allows me to track performance degradations of any given URL,
> so that I can track and focus on adoption perf bottlenecks and opportunities to improve them.

**Acceptance Criteria:**. 
- [x] The plugin must allow users to define a single URL in the plugin options.
- [ ] Run lighthouse against that URL and get audit results in a CP report.
- [x] The plugin accepts all `CLIFlags` as well as `Config` for the official Lighthouse project
- [x] The plugin respects the following flags in relation to the CP data validation:
  - handle audits and groups filtering for `--onlyAudits` and `--onlyCategories`

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR includes all remaining snippets and docs to release the Lighthouse plugin.
In the process of comparing the code with the existing plugins I renamed and moved a couple of files. To get a better overview I grouped the changes as seen below.

**Polish from #549:**
- https://github.com/code-pushup/cli/pull/549#discussion_r1547479530

**Moving existing code:**  
- introduced runner folder to encapsulate lighthouse only related logic and types

**Removed code:**  
  - removed unused `getLighthouseCliArguments` function (better types and direct derivation from lighthouse made it obsolete)

**New code in tests:**  
- e2e tests in cli-e2e
- covered edge case in `getConfig` with error
- added unit tests to `getConfig`
- changed a couple of test names to better readability and narrowed their case (mostly runner utils)

**New code in logic:**  
- added helper functions to create `CategoryRefs`
- docs
- installed dev dependency `@types/debug` for better typing
- I implemented normalisation and refactored the incoming and forwarded `flag` types 
- adjust runtime error message for lighthouse plugin as it was not clear which plugin it came from
Screenshot 2024-03-31 at 15 12 12

> [!NOTE]
> The e2e tests of compare are hard to change as they require a clean history to run. This could get discussed on how to fix it.

**Followup changes:**  
- #598 

closes #90

### CODE REVIEW

File: examples/react-todos-app/code-pushup.config.js [hunk 1]
```diff
@@ -24,9 +29,37 @@ export default {
...
+      slug: 'seo',
+      title: 'SEO',
+      refs: [lighthouseGroupRef('seo')],
+    },
+    {
+      slug: 'pwa',
+      title: 'PWA',
+      isBinary: true,
+      refs: [lighthouseGroupRef('pwa'), eslintAuditRef('react-jsx-key', 1)],
+    },
```
Comment (matejchalk) [inline examples/react-todos-app/code-pushup.config.js:62]: The lint rule is supposed to be in _Performance_ category, not _PWA_:

```suggestion
    {
      slug: 'performance',
      title: 'Performance',
      refs: [lighthouseGroupRef('performance'), eslintAuditRef('react-jsx-key', 1)],
    },
    {
      slug: 'a11y',
      title: 'Accessibility',
      refs: [lighthouseGroupRef('accessibility')],
    },
    {
      slug: 'best-practices',
      title: 'Best Practices',
      refs: [lighthouseGroupRef('best-practices')],
    },
    {
      slug: 'seo',
      title: 'SEO',
      refs: [lighthouseGroupRef('seo')],
    },
    {
      slug: 'pwa',
      title: 'PWA',
      isBinary: true,
      refs: [lighthouseGroupRef('pwa')],
    },
```

File: package.json [hunk 2]
```diff
@@ -82,6 +82,7 @@
     "@trivago/prettier-plugin-sort-imports": "^4.2.0",
     "@types/benchmark": "^2.1.4",
     "@types/chalk": "^2.2.0",
+    "@types/debug": "^4.1.12",
```
Comment (matejchalk) [inline package.json:85]: Are we using the [`debug`](https://github.com/debug-js/debug) package? :raised_eyebrow:

File: e2e/cli-e2e/tests/collect.e2e.test.ts [hunk 3]
```diff
@@ -95,6 +100,21 @@ describe('CLI collect', () => {
     expect(omitVariableReportData(report as Report)).toMatchSnapshot();
   });
 
+  it('should run Lighthouse plugin that runs lighthouse CLI and creates lh-report.json', async () => {
```
Comment (matejchalk) [inline e2e/cli-e2e/tests/collect.e2e.test.ts:103]:
```suggestion
  it('should run Lighthouse plugin that runs lighthouse CLI and creates report.json', async () => {
```

File: packages/plugin-lighthouse/src/lib/runner/constants.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+export const LIGHTHOUSE_NAVIGATION_AUDITS: Audit[] = allRawLighthouseAudits
+  // This plugin only supports the "navigation" mode of Lighthouse in the current implementation
+  // If we don't exclude other audits we throw in the plugin output validation as some of the provided audits are not included in `lighthouse-report.json`
+  .filter(
+    audit =>
+      audit.meta.supportedModes == null ||
+      (Array.isArray(audit.meta.supportedModes) &&
+        audit.meta.supportedModes.includes('navigation')),
+  )
```
Comment (matejchalk) [inline packages/plugin-lighthouse/src/lib/runner/constants.ts:25]: Interesting, thanks the comment which explains why navigation-only :+1:

File: e2e/cli-e2e/tests/__snapshots__/collect.e2e.test.ts.snap [hunk 5]
```diff
@@ -1696,3 +1696,2021 @@ exports[`CLI collect > should run ESLint plugin and create report.json 1`] = `
...
+          "value": 842,
+        },
+        {
+          "description": "Speed Index shows how quickly the contents of a page are visibly populated. [Learn more about the Speed Index metric](https://developer.chrome.com/docs/lighthouse/performance/speed-index/).",
+          "displayValue": "1.0 s",
+          "score": 1,
+          "slug": "speed-index",
+          "title": "Speed Index",
+          "value": 1038,
+        },
```
Comment (matejchalk) [inline e2e/cli-e2e/tests/__snapshots__/collect.e2e.test.ts.snap:1814]: I don't think we can include the `score`, `value` and `displayValue` in the Lighthouse snapshot, it varies too much.

image

File: code-pushup.config.ts [hunk 6]
```diff
@@ -75,15 +79,38 @@ const config: CoreConfig = {
...
     }),
 
-    // see https://github.com/code-pushup/cli/issues/538
-    // await lighthousePlugin({
-    //   url: 'https://staging.code-pushup.dev/login',
-    //   outputPath: join('.code-pushup', LIGHTHOUSE_OUTPUT_FILE_DEFAULT),
-    //   headless: true,
-    // }),
+    await lighthousePlugin('https://codepushup.dev/', {
+      outputPath: join('.code-pushup', LIGHTHOUSE_REPORT_NAME),
```
Comment (matejchalk) [inline code-pushup.config.ts:83]: Why isn't this the default `outputPath` value?

Comment (BioPhoton) [reply code-pushup.config.ts:83]: very goo point!

File: examples/react-todos-app/code-pushup.config.js (hunk 1 above)
Comment (BioPhoton) [reply examples/react-todos-app/code-pushup.config.js]: removed

File: e2e/cli-e2e/tests/__snapshots__/collect.e2e.test.ts.snap (hunk 5 above)
Comment (BioPhoton) [reply e2e/cli-e2e/tests/__snapshots__/collect.e2e.test.ts.snap]: added a helper

File: packages/plugin-lighthouse/src/lib/runner/constants.unit.test.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      'interaction-to-next-paint',
+      'uses-responsive-images-snapshot',
+      'work-during-interaction',
+    ]);
+    expect(
+      LIGHTHOUSE_NAVIGATION_AUDITS.some(({ slug }) =>
+        invalidAuditSlugs.has(slug),
+      ),
+    ).toBe(false);
+  });
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/runner/constants.unit.test.ts:25]: 🧪 One use case per test: Separate invalid audits should be tested separately. These values can be added to the test description via dynamic testing.
```suggestion
    it.each([
    'interaction-to-next-paint',
    'uses-responsive-images-snapshot',
    'work-during-interaction',
  ])(
    'should not include audit %s which does not support the "navigation" mode',
    invalidSlug => {
      expect(
        LIGHTHOUSE_NAVIGATION_AUDITS.every(({ slug }) => slug !== invalidSlug),
      ).toBe(true);
    },
  );
```

File: packages/plugin-lighthouse/src/lib/runner/utils.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        score: score ?? 1, // score can be null
+        value: Number.parseInt(value.toString(), 10),
+        displayValue,
+      };
+
+      if (details == null) {
+        return auditOutput;
+      }
+
+      return auditOutput;
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/runner/utils.ts:59]: ❔ Is the additional branching necessary? The returned value is identical.

File: packages/plugin-lighthouse/src/lib/runner/utils.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    switch (preset) {
+      case 'desktop':
+        return desktopConfig;
+      case 'perf':
+        return perfConfig as Config;
+      case 'experimental':
+        return experimentalConfig as Config;
+      default:
+        // as preset is a string literal the default case here is normally caught by TS and not possible to happen. Now in reality it can happen and preset could be a string not included in the literal.
+        // Therefor we have to use `as string` is used. Otherwise, it will consider preset as type never
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/runner/utils.ts:106]:
```suggestion
        // Therefore we have to use `as string` is used. Otherwise, it will consider preset as type never
```

File: project.json
Comment (Tlacenka) [inline project.json:1]: 🐩 : Please add `plugin-lighthouse` to the code codecov CI/CD workflow in `code-coverage.yml`.

File: packages/plugin-lighthouse/src/lib/runner/constants.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  | 'pwa';
+
+function getMetaString(value: string | IcuMessage): string {
+  if (typeof value === 'string') {
+    return value;
+  }
+  return value.formattedDefault;
+}
+
+async function loadLighthouseAudit(
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/runner/constants.ts:71]: 👩‍💻Maintainability: Ideally, there should only be constants in `constants.ts` file.

File: packages/plugin-lighthouse/README.md [hunk 11]
```diff
@@ -53,7 +53,69 @@ For more infos visit the [official docs](https://developer.chrome.com/docs/light
...
+import { lighthouseAuditRef } from './utils';
+
+export default {
+  // ...
+  categories: [
+    {
+      slug: 'pwa',
+      title: 'PWA',
+      isBinary: true,
+      refs: [lighthouseAuditRef('installable-manifest', 2), lighthouseAuditRef('splash-screen', 1), lighthouseAuditRef('themed-omnibox', 1), lighthouseAuditRef('content-width', 1), lighthouseAuditRef('themed-omnibox', 2), lighthouseAuditRef('viewport', 2), lighthouseAuditRef('maskable-icon', 1), lighthouseAuditRef('pwa-cross-browser', 0), lighthouseAuditRef('pwa-page-transitions', 0), lighthouseAuditRef('pwa-each-page-has-url', 0)],
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/README.md:114]: 📚 Readability: It would be nicer to format the references one per line for better readability.

**Review by Tlacenka — CHANGES_REQUESTED** [review]
Added some feedback below. Most of it should be easy to change.

The main issue I can see is the size of the snapshot - 1600 lines inside one test is not maintainable. Please reduce the number of included audits (and therefore outputs) in the E2E test.

File: packages/plugin-lighthouse/README.md (hunk 11 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/README.md:114]: Considered!

File: packages/plugin-lighthouse/src/lib/runner/constants.ts (hunk 10 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/runner/constants.ts:71]: As this is really only related to the constants and is needed to get the constants I would keep it there.

File: packages/plugin-lighthouse/README.md [hunk 12]
```diff
@@ -53,7 +53,69 @@ For more infos visit the [official docs](https://developer.chrome.com/docs/light
...
+      isBinary: true,
+      refs: [lighthouseGroupRef('pwa')],
+    },
+  ],
+};
+```
+
+#### Reference groups with `lighthouseGroupRef`
+
+The lighthouse categories are reflected as groups.
```
Comment (matejchalk) [inline packages/plugin-lighthouse/README.md:101]:
```suggestion
The Lighthouse categories are reflected as groups.
```

File: packages/plugin-lighthouse/src/lib/constants.ts [hunk 13]
```diff
@@ -1,84 +1,27 @@
...
-  //   shape: string
-  // otherwise it is a JS object maintaining a `path` property
-  //   shape: { path: string, options?: {}; }
-  const path = typeof value === 'string' ? value : value.path;
-  const module = (await import(`lighthouse/core/audits/${path}.js`)) as {
-    default: typeof LHAudit;
-  };
-  return module.default;
+export function isSupportedFlag<T extends string>(flag: T) {
+  return !LIGHTHOUSE_UNSUPPORTED_CLI_FLAGS.has(flag as UnsupportedCliFlags);
```
[the remark this answers is not in the document]
Comment (matejchalk) [reply packages/plugin-lighthouse/src/lib/constants.ts:26]: They're actually unused. I guess the intention was to use them in `validateFlags` function in `utils.ts`, but that code uses the exported `LIGHTHOUSE_UNSUPPORTED_CLI_FLAGS` set directly. 

So these functions should either be removed, or moved to another file - I wouldn't expect a `constants.ts` file to export functions.

File: packages/plugin-lighthouse/src/lib/runner/utils.ts (hunk 8 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/runner/utils.ts:59]: No!

File: e2e/cli-e2e/tests/collect.e2e.test.ts [hunk 14]
```diff
@@ -9,23 +15,45 @@ describe('CLI collect', () => {
...
+  });
   /* eslint-enable @typescript-eslint/no-unused-vars */
 
   beforeEach(async () => {
     await cleanTestFolder('tmp/e2e');
   });
 
+  afterEach(async () => {
+    await rm('.code-pushup', { force: true, recursive: true });
+  });
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply e2e/cli-e2e/tests/collect.e2e.test.ts:55]: Back

File: e2e/cli-e2e/tests/__snapshots__/collect.e2e.test.ts.snap
[the remark this answers is not in the document]
Comment (BioPhoton) [reply e2e/cli-e2e/tests/__snapshots__/collect.e2e.test.ts.snap:1]: adjusted categories and audits listed

File: project.json
Comment (BioPhoton) [reply project.json:1]: Done

File: packages/plugin-lighthouse/README.md (hunk 11 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/README.md:114]: Ah! It is autoformated... :(

File: packages/plugin-lighthouse/src/lib/constants.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/constants.ts]: removed

File: packages/plugin-lighthouse/README.md [hunk 15]
```diff
@@ -53,7 +53,69 @@ For more infos visit the [official docs](https://developer.chrome.com/docs/light
...
+      slug: 'pwa',
+      title: 'PWA',
+      isBinary: true,
+      refs: [lighthouseGroupRef('pwa')],
+    },
+  ],
+};
+```
+
+#### Reference groups with `lighthouseGroupRef`
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/plugin-lighthouse/README.md:99]: done

File: packages/plugin-lighthouse/src/lib/runner/index.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
+export { createRunnerFunction, LighthouseCliFlags } from './runner';
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/runner/index.ts]: It is not exported form the plugin, just into the internals of plugins.

```ts
// packages/plugin-lighthouse/index.ts
import { lighthousePlugin } from './lib/lighthouse-plugin';

export { LIGHTHOUSE_REPORT_NAME } from './lib/runner';
export {
  LIGHTHOUSE_PLUGIN_SLUG,
  LIGHTHOUSE_OUTPUT_PATH,
} from './lib/constants';
export {
  lighthouseAuditRef,
  lighthouseGroupRef,
  LighthouseGroupSlugs,
} from './lib/utils';
export { lighthousePlugin } from './lib/lighthouse-plugin';
export default lighthousePlugin;

```

File: packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts [hunk 17]
```diff
@@ -1,111 +1,77 @@
...
+  const filteredSlugs = new Set(audits.map(({ slug }) => slug));
+  ui().logger.info(
+    `filter: ${JSON.stringify({ skipAudits, onlyAudits, onlyCategories })}`,
+  );
+  ui().logger.info(`audit count: ${JSON.stringify(audits.length)}`);
+  skipAudits.forEach((slug, index) => {
+    ui().logger.info(
+      `Skip audit ${skipAudits.at(
+        index,
+      )} included in audits: ${filteredSlugs.has(slug)}`,
```
Comment (matejchalk) [inline packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts:60]: These logs look nonsensical. 

- there is no context - when running the CLI, it is not at all clear that they come from Lighthouse plugin
  image
- logging all included/excluded audits/categories seems excessive, and the message is completely cryptic
  image

I'm not sure any of this information needs to be logged at all, honestly. They look like debug logs to me.

File: examples/react-todos-app/code-pushup.config.js [hunk 18]
```diff
@@ -24,9 +28,51 @@ export default {
...
+    }),
   ],
   categories: [
-    // TODO: add performance category once Lighthouse plugin implemented, include eslintAuditRef('react-jsx-key', 0)
+    {
+      slug: 'performance',
+      title: 'Performance',
+      refs: [
+        lighthouseGroupRef('performance'),
+        lighthouseAuditRef('largest-contentful-paint'),
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply examples/react-todos-app/code-pushup.config.js:52]: I did it to use also the `lighthouseAuditRef`API. 

I can:
- A) remove it
- B) create a custom category as all audits are also in a group
- C) kepp it as is

I would go for A or C.

File: packages/plugin-lighthouse/src/lib/types.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+      | 'onlyAudits'
+      | 'skipAudits'
+    >
+  >
+> & {
+  onlyGroups?: string | string[];
+} & {
+  onlyAudits?: string | string[];
+  skipAudits?: string | string[];
+};
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/types.ts:29]: I had them separate as onlyGroups is a renamed prop. Next time I will add a comment.

File: packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts (hunk 17 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts:60]: > debug logs
I could use verbose? 

What I will change is:
- make them less cryptic
- add verbose flag
- add plugin name to the msg

Comment (matejchalk) [reply packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts:60]: Sounds good :+1: BTW the logger has a `debug` method, maybe that would be a better fit here than `info`.

File: examples/react-todos-app/code-pushup.config.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (matejchalk) [reply examples/react-todos-app/code-pushup.config.js:52]: I'd go with A, otherwise the example gets convoluted.

File: packages/plugin-lighthouse/src/lib/utils.unit.test.ts [hunk 20]
```diff
@@ -121,684 +112,250 @@ describe('validateOnlyCategories', () => {
...
-      icon: 'file',
-      description: 'Runs test coverage and created audits',
-      audits: filteredAudits,
-      groups: filteredGroups,
-      runner: {
-        command: 'node',
-        outputFile: 'out.json',
-      },
-    };
+    expect(filteredAudits).toHaveLength(1);
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/utils.unit.test.ts:178]: 🧪 Test maintainability: Please rewrite this to include `expect.ObjectContaining` instead of nested property access.
This is applicable to multiple tests.

File: packages/plugin-lighthouse/src/lib/utils.unit.test.ts [hunk 21]
```diff
@@ -121,684 +112,250 @@ describe('validateOnlyCategories', () => {
 });
 
 describe('filterAuditsAndGroupsByOnlyOptions to be used in plugin config', () => {
-  it('should return given audits and groups if no only filter is set', () => {
-    const audits: Audit[] = [{ slug: 'speed-index', title: 'Speed Index' }];
-    const groups: Group[] = [
+  type PartialGroup = Partial<
+    Omit<Group, 'refs'> & { refs: Partial<Group['refs'][number]>[] }
+  >;
+  const basePluginConfig = (
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/utils.unit.test.ts:118]: 🧪 Maintainability: Is it possible to create tests with minimum audits or groups to avoid this need of abstraction and additional logic? What if each test only has 0-1 audit and 0-1 group?

File: packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts (hunk 17 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/lighthouse-plugin.ts]: Done

File: packages/utils/src/lib/file-system.ts [hunk 22]
```diff
@@ -39,7 +39,7 @@ export async function ensureDirectoryExists(baseDir: string) {
     await mkdir(baseDir, { recursive: true });
     return;
   } catch (error) {
-    ui().logger.error((error as { code: string; message: string }).message);
+    ui().logger.info((error as { code: string; message: string }).message);
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/src/lib/file-system.ts:42]: the process exits otherwise. It is not just the color. Same fro warnings.

File: e2e/cli-e2e/tests/collect.e2e.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (BioPhoton) [reply e2e/cli-e2e/tests/collect.e2e.test.ts:55]: Ah you mean i should delete this code! OK. will do!

File: examples/react-todos-app/code-pushup.config.js [hunk 23]
```diff
@@ -24,9 +28,51 @@ export default {
       eslintrc: '.eslintrc.js',
       patterns: ['src/**/*.js', 'src/**/*.jsx'],
     }),
+    await lighthousePlugin('https://codepushup.dev/', {
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply examples/react-todos-app/code-pushup.config.js:31]: It requires lots of extra effort and the tests are way slower. In the future I can work on a offline test but not for now.

File: packages/utils/src/lib/types.ts [hunk 24]
```diff
@@ -0,0 +1,3 @@
+export type ExcludeNullFromPropertyTypes<T> = {
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/src/lib/types.ts:1]: Did not manage to achieve it with this helper. Tried it before.

File: packages/plugin-lighthouse/src/lib/utils.unit.test.ts (hunk 21 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/utils.unit.test.ts:118]: The schema parsing was involved. so I had to pass valid shapes.

File: e2e/cli-e2e/tests/collect.e2e.test.ts [hunk 25]
```diff
@@ -9,23 +15,45 @@ describe('CLI collect', () => {
   const exampleAuditTitle = 'Disallow unused variables';
 
   /* eslint-disable @typescript-eslint/no-unused-vars */
-  const omitVariableData = ({
+  const omitVariableAuditData = ({
+    score,
+    value,
+    displayValue,
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply e2e/cli-e2e/tests/collect.e2e.test.ts:21]: I was asked by @matejchalk  to remove the data as it was variable. In the future we will implement audit details and there we will write lots of checks :)

**Review by matejchalk — APPROVED** [review]

File: packages/plugin-lighthouse/src/lib/utils.unit.test.ts (hunk 20 above)
Comment (BioPhoton) [reply packages/plugin-lighthouse/src/lib/utils.unit.test.ts:178]: I had that before and did it by following your last suggestion of testing each line witch a separate expect. :)

**Review by matejchalk — APPROVED** [review]

File: e2e/cli-e2e/tests/collect.e2e.test.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (Tlacenka) [reply e2e/cli-e2e/tests/collect.e2e.test.ts:20]: I understand. However, in the past only some of the results were omitted but by looking at the omitted values for the Lighthouse plugin specifically, **all results** are omitted. So the E2E test only checks if the plugin runs, not that it returns correct results.

File: packages/plugin-lighthouse/src/lib/utils.unit.test.ts [hunk 26]
```diff
@@ -121,684 +112,250 @@ describe('validateOnlyCategories', () => {
...
-    };
+    expect(filteredAudits).toHaveLength(1);
+    expect(filteredAudits.at(0)?.slug).toBe('first-contentful-paint');
+
+    expect(filteredGroups).toHaveLength(1);
+    expect(filteredGroups.at(0)?.slug).toBe('performance');
+    expect(filteredGroups.at(0)?.refs).toHaveLength(1);
+    expect(filteredGroups.at(0)?.refs.at(0)?.slug).toBe(
+      'first-contentful-paint',
+    );
```
Comment (Tlacenka) [inline packages/plugin-lighthouse/src/lib/utils.unit.test.ts:186]: I do not recall advocating for nested conditional chaining in tests where the object structure should only have one expected value. If you can find the comment, please send it.

The approach below cleanly avoids the TypeScript complaints. Same approach to be used in other applicable cases.

```suggestion
    expect(filteredAudits).toStrictEqual([{ slug: 'first-contentful-paint' }]);
    expect(filteredGroups).toStrictEqual([
      {
        slug: 'performance',
        refs: [{ slug: 'first-contentful-paint' }],
      },
    ]);
```

**Review by Tlacenka — APPROVED** [review]
The last commented changes can be done in a follow-up PR.

File: e2e/cli-e2e/tests/collect.e2e.test.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (Tlacenka) [reply e2e/cli-e2e/tests/collect.e2e.test.ts:20]: Off the top of my head, accessibility or cumulative layout shift metrics should not be as dynamic if we choose a FE application under our control. So their results could be used in E2E tests.

**Comment by BioPhoton** [comment]
The following comments will get addressed in a followup PR:
- https://github.com/code-pushup/cli/pull/593#discussion_r1559788572
- https://github.com/code-pushup/cli/pull/593#discussion_r1559803319

File: e2e/cli-e2e/tests/collect.e2e.test.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (BioPhoton) [reply e2e/cli-e2e/tests/collect.e2e.test.ts:20]: I get your point and noted it as TODO. It is not that easy to list all no variable audits and it would add additional maintenance effort, but I'll stat to list non variable audits in a ticket.
